deeplearning4j-nn/src/main/java/org/deeplearning4j/nn/layers/convolution/Deconvolution2DLayer.java

```diff
@@ -0,0 +1,288 @@
+
+        INDArray outEpsilon = Nd4j.create(miniBatch * inDepth * inH * inW);
+        INDArray reshapedEpsilon = outEpsilon.reshape('c', miniBatch, inDepth, inH, inW);
+
+        Integer sameMode = (convolutionMode == ConvolutionMode.Same) ? 1 : 0;
+
+        int[] args = new int[] {
+                kH, kW, strides[0], strides[1],
+                pad[0], pad[1], dilation[0], dilation[1], sameMode
+        };
+
+        INDArray delta;
+        IActivation afn = layerConf().getActivationFn();
+        Pair<INDArray, INDArray> p = preOutput4d(true, true);
+        delta = afn.backprop(p.getFirst(), epsilon).getFirst();
+
+        CustomOp op;
+        if(layerConf().hasBias()){
+
+            bias = getParamWithNoise(DeconvolutionParamInitializer.BIAS_KEY, true);
+
+            op = DynamicCustomOp.builder("deconv2d_bp")
+                    .addInputs(input, weights, bias, delta)
+                    .addIntegerArguments(args)
+                    .addOutputs(reshapedEpsilon, weightGradView, biasGradView)
+                    .callInplace(false)
+                    .build();
+        } else {
+            op = DynamicCustomOp.builder("deconv2d_bp")
+                    .addInputs(input, weights, delta)
+                    .addIntegerArguments(args)
+                    .addOutputs(reshapedEpsilon, weightGradView)
+                    .callInplace(false)
+                    .build();
+        }
+        Nd4j.getExecutioner().exec(op);
+
+        Gradient retGradient = new DefaultGradient();
+        if(layerConf().hasBias()){
+            retGradient.setGradientFor(DeconvolutionParamInitializer.BIAS_KEY, biasGradView);
+        }
+        retGradient.setGradientFor(DeconvolutionParamInitializer.WEIGHT_KEY, weightGradView, 'c');
+        weightNoiseParams.clear();
+
+        reshapedEpsilon = reshapedEpsilon.permute(1, 0, 2 , 3);
+
+        return new Pair<>(retGradient, reshapedEpsilon);
+    }
+
+
+    @Override
+    public INDArray preOutput(boolean training) {
+        return preOutput(training, false).getFirst();
+    }
+
+    protected Pair<INDArray, INDArray> preOutput(boolean training , boolean forBackprop) {
+
+        if (convolutionMode == ConvolutionMode.Same) {
+            throw new IllegalArgumentException("Border mode Same currently not supported.");
+        }
+
+        INDArray bias = getParamWithNoise(DeconvolutionParamInitializer.BIAS_KEY, training);
+        INDArray weights = getParamWithNoise(DeconvolutionParamInitializer.WEIGHT_KEY, training);
+
+        //Input validation: expect rank 4 matrix
+        if (input.rank() != 4) {
+            String layerName = conf.getLayer().getLayerName();
+            if (layerName == null)
+                layerName = "(not named)";
+            throw new DL4JInvalidInputException("Got rank " + input.rank()
+                    + " array as input to SeparableConvolution2D (layer name = " + layerName + ", layer index = "
```

Comment:
Wrong layer name (should be deconv, not separable)

Reply:
good catch! :)